qiskit/algorithms/eigen_solvers/vqd.py

```diff
@@ -0,0 +1,840 @@
+                    result.optimal_point,
+                    self._eval_count,
+                )
+            else:
+                logger.info(
+                    (
+                        "%s excited state optimization complete in %s s.\nFound opt_parms %s in %s evals"
+                    ),
+                    str(step - 1),
+                    eval_time,
+                    result.optimal_point,
+                    self._eval_count,
+                )
+
+        # To match the siignature of NumpyEigenSolver Result
+        result.eigenstates = ListOp([StateFn(vec) for vec in result.eigenstates])
+        result.eigenvalues = np.array(result.eigenvalues)
+        result.optimal_point = np.array(result.optimal_point)
+        result.optimal_value = np.array(result.optimal_value)
+        result.cost_function_evals = np.array(result.cost_function_evals)
+        result.optimizer_time = np.array(result.optimizer_time)
+
+        # TODO delete as soon as get_optimal_vector etc are removed
+        self._ret = result
+
+        if aux_operators is not None:
+            result.aux_operator_eigenvalues = aux_values
+
+        return result
+
+    def get_energy_evaluation(
+        self,
+        step: int,
+        operator: OperatorBase,
+        return_expectation: bool = False,
+        prev_states: List[float] = None,
+    ) -> Callable[[np.ndarray], Union[float, List[float]]]:
+        """Returns a function handle to evaluates the energy at given parameters for the ansatz.
+
+        This is the objective function to be passed to the optimizer that is used for evaluation.
+
+        Args:
+            step: level of enegy being calculated. 0 for ground, 1 for first excited state and so on.
+            operator: The operator whose energy to evaluate.
+            return_expectation: If True, return the ``ExpectationBase`` expectation converter used
+                in the construction of the expectation value. Useful e.g. to evaluate other
+                operators with the same expectation value converter.
+            prev_states: List of parameters from previous rounds of optimization.
+
+
+        Returns:
+            Energy of the hamiltonian of each parameter, and, optionally, the expectation
+            converter.
+
+        Raises:
+            RuntimeError: If the circuit is not parameterized (i.e. has 0 free parameters).
+            AlgorithmError: If operator was not provided.
+
+        """
+
+        num_parameters = self.ansatz.num_parameters
+        if num_parameters == 0:
+            raise RuntimeError("The ansatz must be parameterized, but has 0 free parameters.")
+
+        if operator is None:
+            raise AlgorithmError("The operator was never provided.")
+
+        if step > 1 and (len(prev_states) + 1) != step:
+            raise RuntimeError(
+                "Passed previous states of the wrong size. Passed array has length "
+                + str(len(prev_states))
```

Comment:
You should be able to use an f"...." string here i.e. `f"Passed previous states of the wrong size. Passed array has length {len(prev_states)}"` Maybe this is better than an indexing error which would occur otherwise but normally its only VQD that calls this right so in this case it really is a coding/internal error.